Adds Vali chemical enhancement module and Harvester blade

Walarks opened the pull request:
<!-- ***STOP!***  Read this: If this is not a PR ready for review and merge or WIP, open it as a draft PR, using the arrow next to 'Create Pull Request'>

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Adds the Vali Chemical Enhancement module and harvester blade. This loadout's playstyle is centered around using reagents and getting in melee range rather than using guns from range and carrying ammo.

The system stores "green blood" that is collected when you melee a living xeno with a connected compatible weapon. Then it uses the stored resource to enhance the effectiveness of reagents in you. It is intended to have lower armor in exchange of increased survivability through the use of reagent effects.

Vali module:
It has a cooldown of 10s after being deactivated
An overlay appears when held for more than 12s
Causes necrotising flesh on a random limb if held active for more than 20s and an additional limb for every 10 seconds past that
200 green blood storage
Drains 4 or 12 green blood per 2 seconds
Has 2 tiers of boosts - 2x and 4x reagent effectiveness
Can link the Harvester blade to itself. The blade gains nodrop and gives the system 20u of green blood per attack on a living beno.
Can extract green blood from itself into a held beaker.

Harvester blade:
- 60 damage (like the machete)
- Can be loaded with bicaridine/kelotane/tramadol to empower your next attack. Can contain only a single kind of reagent at the same time.
- Stores up to 40u. Filled using a reagent container and emptied using an empty beaker
To use the effect, the weapon needs to be loaded with up to 40u of a compatible reagent from a container that contains only one type of reagent. Then it takes 2 seconds to load a single use of the effect for 10u, during which you can walk with 75% slowdown.

- Effects:
1. Bicaridine
Your next attack instead channels for 2 seconds and heals 10 brute on the target
2. Kelotane
You produce a flame cone for 2 tiles in the direction you slash
3. Tramadol
You apply 2 slowdown for 2 seconds on the target

Virilyth(green blood) reagent:
- Can react with bicaridine/kelotane/tramadol/dylovene to produce 1u of it for 1u of green blood used.

Also added code for general generation of cones.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Adds a combat option for people who want to heavily opt into using chems.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Added Vali chemical enhancement module and harvester blade
add: Added generalized cone generation code
add: Added virilyth reagent. It reacts with dylovene/bicaridine/kelotane/tramadol and duplicates them in a 1:1 ratio. Can be acquired by extracting green blood from the Vali module
tweak: added signal for when a reagent is added for the first time in a holder
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

tgstation-server commented:
You currently have a negative Fix/Feature pull request delta of -18. Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.

neohenry commented:
MELEE SPEC INCOMING

SilencedMP5A5 commented:
pog

revelation-8468 commented:
BERSERKER MARINE

David-Stormwell commented:
this will be fun to watch

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+	COOLDOWN_DECLARE(chemboost_activation_cooldown)
+
+/datum/component/chem_booster/Initialize()
+	if(!istype(parent, /obj/item))
```
Isitem

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+/**
+	Chem booster component
+
+	This component stores "green blood" and uses it to increase effect_str of chems in the user's body.
+
```
Autodoc can work on types too so make this doc proper if you don't mind

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+
+	///Amount of substance that the component can store
+	var/resource_storage_max = 200
+	///Ammount of sunstance stored currently
```
```suggestion
	///Amount of sunstance stored currently
```
Substance is what you mean here?

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+	resource_storage_current += min(resource_storage_max - resource_storage_current, 20)
+
+/datum/action/chem_booster/configure
+	name = "Configure Vali CHemical Enhancement"
```
Typo?

Walarks replied:
changed

TiviPlus commented on code/game/objects/items/weapons/blades.dm:
```diff
@@ -20,6 +20,130 @@
+	force = 60
+	attack_speed = 12
+	w_class = WEIGHT_CLASS_BULKY
+	var/obj/item/reagent_containers/glass/beaker/vial/beaker = null
+	flags_item = DRAINS_XENO
+
```
Declare new vars after redefining old ones

Walarks replied:
changed

TiviPlus commented on code/game/objects/items/weapons/blades.dm:
```diff
@@ -20,6 +20,130 @@
+	)
+
+/obj/item/weapon/claymore/harvester/examine(mob/user)
+	..()
```
. = ..()

Walarks replied:
changed

TiviPlus commented on code/game/objects/items/weapons/blades.dm:
```diff
@@ -20,6 +20,130 @@
+				var/turf/target = get_turf(M)
+				target.ignite()
+				var/list/cone_turfs = generate_cone(user, 2, 1, 91, Get_Angle(user, M.loc))
+				for(var/turf/T in cone_turfs)
```
For loop that type checks when it shouldnt

Walarks replied:
I get "code\game\objects\items\weapons\blades.dm:131:error: T.ignite: undefined proc" when I just use var/T instead of var/turf/T.

Walarks replied:
changed

TiviPlus commented on code/game/objects/items/weapons/blades.dm:
```diff
@@ -20,6 +20,130 @@
+
+			if(/datum/reagent/medicine/bicaridine)
+				var/target_self = TRUE
+				if(!M == user)
```
Uh?

Walarks replied:
removed this part

TiviPlus commented on code/modules/reagents/reagents/other.dm:
```diff
@@ -658,3 +658,19 @@
+
+/*
+/datum/reagent/virilyth
+	name = "Virilyth"
+	description = "A strange substance with high regenerative properties."
+	color = "#488C54"
+	taste_description = "slimy lemon"
+	can_synth = FALSE
```
Gonna assume those is wip

TiviPlus replied:
this is still commented out for some reason?

Walarks replied:
Will either remove it for a later PR or add it with a way to extract green blood from the suit and mix it with certain common reagents to duplicate them.

Depends on how the meds economy goes during the testmerge.

TiviPlus commented on code/datums/status_effects/debuffs.dm:
```diff
@@ -228,3 +228,19 @@
+		return
+	owner.add_movespeed_modifier(MOVESPEED_ID_HARVEST_TRAM_SLOWDOWN, TRUE, 0, NONE, TRUE, debuff_slowdown)
+
+/datum/status_effect/noplasmaregen/on_remove()
```
Copypaste

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+		message_admins("added chem boosted by > [boost_amount]")
+		R.effect_str += boost_amount
+		break
+
+/datum/component/chem_booster/proc/connect_weapon()
```
Reminder new procs need to be autodocced

Walarks replied:
changed

TiviPlus commented on code/__HELPERS/patterns.dm:
```diff
@@ -79,3 +79,38 @@
 	return results
+
+///Generates a cone shape. Any other checks should be handled with the resulting list. Can work with up to 359 degrees
+/proc/generate_cone(atom/center, outer_range = 10, inner_range = 1, cone_width = 60, cone_direction = 0, blocked = TRUE)
```
Reminder to self

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+/datum/component/chem_booster/Initialize()
+	if(!istype(parent, /obj/item))
+		return COMPONENT_INCOMPATIBLE
+	boost_amount = 0
```
?

Walarks replied:
removed boost_amount

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+	configure_action.give_action(wearer)
+
+/datum/component/chem_booster/process()
+	if(resource_storage_current < resource_drain_amount)
+		to_chat(wearer, "<span class='warning'>Insufficient resources to maintain operation.</span>")
+		on_off()
+	resource_storage_current -= resource_drain_amount
+
```
Could this not make it negative

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,210 @@
+ *	Opens the radial menu with everything
+ */
+/datum/component/chem_booster/proc/configure(datum/source)
+	SIGNAL_HANDLER_DOES_SLEEP
```
New signal handler does sleep are not permitted

Walarks replied:
removed this line

TiviPlus reviewed (changes requested):
Surface review since it's wip

Tennessee116 commented:
So this is a positive feedback loop?

Walarks commented:
> So this is a positive feedback loop?

Yeah, with increasing penalties for using it too much.

Walarks commented:
Feature-wise I think it is good enough for testmerge to see if the concept works and I should polish it more.

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+ *
+ */
+/datum/component/chem_booster
+	var/mob/living/carbon/human/wearer
+
```
Is there a particular reason you use hard refs instead of signal args to get a wearer reference here?

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+	if(!weapon_to_connect)
+		return FALSE
+
+	connected_weapon = weapon_to_connect
+	ENABLE_BITFIELD(connected_weapon.flags_item, NODROP)
+	RegisterSignal(connected_weapon, COMSIG_ITEM_ATTACK, .proc/drain_resource)
```
No hard del protection

Walarks replied:
I'm not sure what you mean here.

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+/**
+ *	Opens the radial menu with everything
+ */
+/datum/component/chem_booster/proc/configure(datum/source)
+	var/mob/living/carbon/human/H = wearer
+
```
No signal handler

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+//Updates the effect_str of chems that enter the body while the component is on
+/datum/component/chem_booster/proc/late_add_chem(datum/source, datum/reagent/added_chem, amount)
+	SIGNAL_HANDLER
+	for(var/datum/reagent/R in wearer.reagents.reagent_list)
+		if(!istype(R, added_chem))
+			continue
```
Istyping list

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+	RegisterSignal(wearer.reagents, COMSIG_REAGENT_ADD, .proc/late_add_chem)
+	RegisterSignal(wearer, COMSIG_MOB_DEATH, .proc/on_off)
+	update_boost(boost_amount*2, FALSE)
+	to_chat(wearer, "Commensing reagent injection.")
+
```
Add some spans

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+		wearer.clear_fullscreen("degeneration")
+		var/necrotized_counter = FLOOR(min(world.time-processing_start, 20 SECONDS)/200 + (world.time-processing_start-20 SECONDS)/100, 1)
+		if(necrotized_counter >= 1)
+			for(var/datum/limb/L in shuffle(wearer.limbs))
+				if(L.limb_status & LIMB_NECROTIZED)
```
typechecking list

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+/datum/component/chem_booster/proc/configure(datum/source)
+	var/mob/living/carbon/human/H = wearer
+
+	var/list/radial_options = list(
+		"on_off" = image(icon = 'icons/mob/radial.dmi', icon_state = "cboost_on_off"),
+		"connect" = image(icon = 'icons/mob/radial.dmi', icon_state = "cboost_connect"),
+		"boost" = image(icon = 'icons/mob/radial.dmi', icon_state = "[boost_icon]"),
+		)
```
Is there a reason this isnt a static list?

Walarks replied:
changed

TiviPlus commented on code/__HELPERS/patterns.dm:
```diff
@@ -79,3 +79,38 @@
 				results += AdjT
 	return results
+
+///Generates a cone shape. Any other checks should be handled with the resulting list. Can work with up to 359 degrees
+/proc/generate_cone(atom/center, outer_range = 10, inner_range = 1, cone_width = 60, cone_direction = 0, blocked = TRUE)
```
Could you make this documentation more clear as to what the params do here?

Walarks replied:
changed

TiviPlus commented on code/__HELPERS/patterns.dm:
```diff
@@ -79,3 +79,38 @@
+	var/list/turfs_to_check = list(get_turf(center))
+	var/list/cone_turfs = list()
+	for(var/r in 1 to outer_range)
+		for(var/turf/C in turfs_to_check)
```
typechecking list

Walarks replied:
changed

TiviPlus commented on code/__HELPERS/patterns.dm:
```diff
@@ -79,3 +79,38 @@
+				if(turf_angle > right_angle && turf_angle < left_angle)
+					continue
+				if(blocked)
+					if(T.density || LinkBlocked(C, T) || TurfBlockedNonWindow(T))
+						continue
```
Doesnt linkblocked check density already?
And turfblockednonwindow will never be called because it also contains a density check

Walarks replied:
I tried removing the T.density check, but it includes tiles with full walls if I do that

TiviPlus commented on code/__HELPERS/patterns.dm:
```diff
@@ -79,3 +79,38 @@
+	if(left_angle < 0)
+		left_angle += 360
+
+	var/list/turfs_to_check = list(get_turf(center))
```
Ca you rename this with rowtocheck  or similar or something for clarity

Walarks replied:
renamed to max_row_count

TiviPlus commented on code/__HELPERS/patterns.dm:
```diff
@@ -79,3 +79,38 @@
+	var/list/cone_turfs = list()
+	for(var/r in 1 to outer_range)
+		for(var/turf/C in turfs_to_check)
+			for(var/direction in GLOB.cardinals)
+				var/turf/T = get_step(C, direction)
+				if(cone_turfs.Find(T))
+					continue
+				if(get_dist(center, T) > outer_range || get_dist(center, T) < inner_range)
+					continue
+				var/turf_angle = Get_Angle(center, T)
```
This .find seems very strange 
you also shouldnt really be checking all the dirs, that seems very wasteful, consider excluding the reverse dir with REVERSE_DIR and or the diagonal dirs as well

additionally is the
`get_dist(center, T) > outer_range`
really needed seeing as you to for(x to outer_range) already

Walarks replied:
Removed get_dist(center, T) > outer_range, along with removing the opposite direction. The proc otherwise doesn't check the diagonal directions.

Also moved the proc since REVERTSE_DIR was defined after the proc in it previous spot.

Walarks commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,237 @@
+		boost_amount += amount
+	resource_drain_amount = boost_amount*4
+	if(boost_on)
+		for(var/datum/reagent/R in wearer.reagents.reagent_list)
+			R.effect_str += amount
```
changed

Walarks commented on code/game/objects/items/weapons/blades.dm:
```diff
@@ -20,6 +20,128 @@
+	. = ..()
+	var/list/traits = list()
+
+	traits += "Tramadol - slow your target"
+	traits += "Kelotane - produce a cone of flames"
+	traits += "Bicaridine - heal your target"
+
+	. += jointext(traits, "<br>")
```
changed

Walarks replied:
reverted this as using \n doesn't make the parts appear on multiple lines

Walarks commented on code/game/objects/items/weapons/blades.dm:
```diff
@@ -20,6 +20,128 @@
+	loaded_reagent = beaker.reagents.reagent_list[1].type
+	return TRUE
+
+/obj/item/weapon/claymore/harvester/attack(mob/living/M, mob/living/user)
+	if(effect_active)
+		switch(loaded_reagent)
+			if(/datum/reagent/medicine/tramadol)
```
changed

Walarks commented on code/game/objects/machinery/vending/marine_vending.dm:
```diff
@@ -856,6 +856,8 @@
 
 	product_ads = "If it moves, it's hostile!;How many enemies have you killed today?;Shoot first, perform autopsy later!;Your ammo is right here.;Guns!;Die, scumbag!;Don't shoot me bro!;Shoot them, bro.;Why not have a donut?"
 	products = list(
+		/obj/item/clothing/suit/storage/marine/chemsuit = 10,
+		/obj/item/weapon/claymore/harvester = 15,
```
I added it here because people I talked to on discord feel it is on the weak side and this would be a good place to add it for testing. Though if you like to I'll move it to req for 15 points or something like this. 

Also the question about using the suit+weapon requiring expending of a large quantity of meds to work was brought up, so req could spend points on that instead.

Walarks commented on code/modules/reagents/holder.dm:
```diff
@@ -609,6 +609,7 @@
 		my_atom.on_reagent_change(ADD_REAGENT)
 	if(!no_react)
 		handle_reactions()
+	SEND_SIGNAL(src, COMSIG_REAGENT_ADD, reagent, amount)
 	return TRUE
```
I added it there because the same proc handles both adding new and existing reagents, while I only need new ones.

TiviPlus commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1372,3 +1372,52 @@ will handle it, but:
 			if(A.type == areatype)
 				areas += V
 	return areas
+
+/*
```
```suggestion

/**
```

InterroLouis commented:
It needs a suit light.

Walarks commented:
> It needs a suit light.

Added just now.

InterroLouis commented:
My thoughts are, perhaps a little more melee armor, not a necessity but it would help the melee only person to be able to function batter. 
Second, a health analyzer, and while I would _sort of_ like autosplint that probably isn't going to fly. So just the health analyzer, since it's all chem focused and knowing what chems are in you is pretty important to that. 
Third, maybe a chem storage pool in the suit instead of the sword if that can be made to work. 
Four, I'd kind of like to see what chems it can use be increased, maybe with some more niche uses for certain chems. It would make the kelotane swing less dangerous to the user(fire bad unga dunga) if you had it just ignite the target instead of making a bunch of fire on the ground.

David-Stormwell commented:
I'm going to suggest a quicker way to unconnect and reconnect the sword, in case you are being gunned down by a Spitter and can't do anything about it, etc.
Also, if the jetpack PR didn't exist I'd suggest jump-boots of some sort, or something of that case.

InterroLouis commented:
Also I want to be able to whack the shield with the sword to make a thump thump thump sound.

XSlayer300 commented on code/game/objects/items/weapons/blades.dm:
```diff
@@ -20,6 +20,143 @@
 	return(BRUTELOSS)
 
+/obj/item/weapon/claymore/harvester
+	name = "\improper Harvester"
```
```suggestion
	name = "\improper HP-S Harvester blade"
```
Gotta give it a fancy name.

XSlayer300 commented on code/modules/clothing/suits/marine_armor.dm:
```diff
@@ -343,6 +343,25 @@
 	icon_state = "marine_sniperm"
 
+/obj/item/clothing/suit/storage/marine/chemsuit
+	name = "Vali Chemical Enhancement suit"
```
```suggestion
	name = "\improper Vali chemical enhancement suit"
```

Walarks commented:
> My thoughts are, perhaps a little more melee armor, not a necessity but it would help the melee only person to be able to function batter.
> Second, a health analyzer, and while I would _sort of_ like autosplint that probably isn't going to fly. So just the health analyzer, since it's all chem focused and knowing what chems are in you is pretty important to that.
> Third, maybe a chem storage pool in the suit instead of the sword if that can be made to work.
> Four, I'd kind of like to see what chems it can use be increased, maybe with some more niche uses for certain chems. It would make the kelotane swing less dangerous to the user(fire bad unga dunga) if you had it just ignite the target instead of making a bunch of fire on the ground.

Not sure about the armor. I want to try playing a bit more around having survivability be based on chems, though I may give more armor for the sake of overall simplicity later.

Added a health analyzer like valk's now, though may make it to only show chems vs the detailed scan of normal analyzers.

Not sure about a better way to change the mechanics behind selecting harvester's reagents yet. If I were to increase the internal storage, I would look into decreasing costs instead. I removed the need to prime the effect a second time before being able to use it on someone, so this should make it a lot less clunky I think.

If you have any ideas for effects for other chems, I can look into adding them or changing existing ones. I may change kelotane's effect to only proc for one tile, but I am thinking of having tramadol be the more spammable reagent and kelotane slower with higher impact, so I may make changes around this idea.

Walarks commented:
> I'm going to suggest a quicker way to unconnect and reconnect the sword, in case you are being gunned down by a Spitter and can't do anything about it, etc.
> Also, if the jetpack PR didn't exist I'd suggest jump-boots of some sort, or something of that case.

I decreased the time it takes to connect the sword from 4s to 2s.

Walarks commented:
> Also I want to be able to whack the shield with the sword to make a thump thump thump sound.

This could be added in a different PR since it is out of the area of this one as other swords can't do it either.

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+	QDEL_NULL(power_action)
+	QDEL_NULL(scan_action)
+
+/datum/component/chem_booster/proc/examine(datum/source, mob/user)
+	SIGNAL_HANDLER
```
Autodoc

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+	SIGNAL_HANDLER
+	to_chat(user, "<span class='notice'>The chemical system currently holds [resource_storage_current]u of green blood. Its' enhancement level is set to [boost_amount+1].</span>")
+
+/datum/component/chem_booster/proc/dropped(datum/source, mob/user)
+	SIGNAL_HANDLER
```
Autodoc

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+	power_action.remove_action(wearer)
+	scan_action.remove_action(wearer)
+	wearer = null
+
+/datum/component/chem_booster/proc/equipped(datum/source, mob/equipper, slot)
```
Autodoc

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+
+		if("extract")
+			extract(10)
+
+/datum/component/chem_booster/proc/on_off(datum/source)
```
Autodoc

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+			continue
+		var/datum/reagent/R = X
+		R.effect_str += boost_amount
+		break
+
+/datum/component/chem_booster/proc/scan_user(datum/source)
```
Autodoc

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+
+/datum/component/chem_booster/proc/scan_user(datum/source)
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, .proc/act_scanner)
+
+datum/component/chem_booster/proc/act_scanner()
+	analyzer.attack(wearer, wearer, TRUE)
```
just invoke async in scan_user

Walarks replied:
changed

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+	to_chat(wearer, "<span class='notice'>You finish connecting the [held_item].</span>")
+	wearer.remove_movespeed_modifier(MOVESPEED_ID_CHEM_CONNECT)
+	manage_weapon_connection(held_item)
+
+/datum/component/chem_booster/proc/manage_weapon_connection(obj/item/weapon_to_connect)
```
Autodoc

Walarks replied:
changed

TiviPlus commented on code/modules/clothing/suits/marine_armor.dm:
```diff
@@ -343,6 +343,35 @@
+
+/obj/item/clothing/suit/storage/marine/chemsuit/apply_custom(image/standing)
+	. = ..()
+	message_admins("signal test")
```
.

Walarks replied:
removed message_admins

TiviPlus commented on code/modules/clothing/suits/marine_armor.dm:
```diff
@@ -343,6 +343,35 @@
+/obj/item/clothing/suit/storage/marine/chemsuit/apply_custom(image/standing)
+	. = ..()
+	message_admins("signal test")
+	var/image/resource_amount = image('icons/mob/hud.dmi',icon_state = "chemsuit_vis")
+	var/datum/component/chem_booster/C = datum_components[/datum/component/chem_booster]
+	var/resource_percent = 	C.resource_storage_current/C.resource_storage_max
+	resource_amount.color = rgb(61, 185, 4, 255*resource_percent)
+	standing.overlays += resource_amount
```
is this some sort of hacked together getcomponent?
this should likely  be using signals instead of whatever this is

Walarks replied:
changed

TiviPlus commented on code/__DEFINES/dcs/signals.dm:
```diff
@@ -220,6 +220,11 @@
 
+#define COMSIG_NEW_REAGENT_ADD	"new_reagent_add"					//from add_reagent(): (/datum/reagent, amount); it is sent when a reagent gets added for the first time to a holder
+
+#define COMSIG_CHEM_BOOSTER_ON_OFF "chem_booster_on_off"		//from base of on_off(): (/datum/component/chem_booster, boost_on)
```
I dont see this listened for anywhere

Walarks replied:
removed COMSIG_CHEM_BOOSTER_ON_OFF signal

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+
+	if(new_resource_str_tier != current_resource_str_tier)
+		var/resource_percentage = (resource_storage_current+amount_added)/resource_storage_max
+		SEND_SIGNAL(parent, COMSIG_CHEM_BOOSTER_RES_UPD, resource_percentage, amount_added)
```
Why not use parent here?

Walarks replied:
changed, removed signal

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+	to_chat(wearer, "<span class='notice'>Commensing reagent injection.</span>")
+	playsound(get_turf(wearer), 'sound/effects/bubbles.ogg', 30, 1)
+
+//Updates the boost amount of the suit and effect_str of reagents if component is on. "amount" is the final level you want to set the boost to.
```
```suggestion
///Updates the boost amount of the suit and effect_str of reagents if component is on. "amount" is the final level you want to set the boost to.
```

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+			var/datum/reagent/R = X
+			R.effect_str += amount
+
+//Updates the effect_str of chems that enter the body while the component is on
```
```suggestion
///Updates the effect_str of chems that enter the body while the component is on
```

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,330 @@
+	RegisterSignal(connected_weapon, list(COMSIG_ITEM_EQUIPPED_NOT_IN_SLOT, COMSIG_ITEM_DROPPED), .proc/connect_weapon)
+	return TRUE
+
+//Handles resource collection and is ativated when attacking with a weapon.
```
```suggestion
///Handles resource collection and is ativated when attacking with a weapon.
```

TiviPlus commented on code/datums/components/chem_booster.dm:
```diff
@@ -0,0 +1,334 @@
+	var/obj/item/connected_weapon
+	///When was the effect activated. Used to activate negative effects after a certain amount of use
+	var/processing_start = 0
+	var/static/image/resource_overlay = image('icons/mob/hud.dmi', icon_state = "chemsuit_vis")
```
Autodoc

Walarks replied:
changed

Walarks replied:
changeds

TiviPlus commented on code/game/objects/machinery/vending/new_marine_vendors.dm:
```diff
@@ -365,6 +365,7 @@ GLOBAL_LIST_INIT(marine_selector_cats, list(
 		/obj/effect/essentials_set/mimir = list(CAT_ARMMOD, "Mark 1 Mimir Resistance set", 0,"black"),
 		/obj/item/armor_module/attachable/tyr_extra_armor/mark1 = list(CAT_ARMMOD, "Mark 1 Tyr extra armor module", 0,"black"),
 		/obj/item/armor_module/attachable/better_shoulder_lamp/mark1 = list(CAT_ARMMOD, "Mark 1 Baldur light armor module", 0,"black"),
+		/obj/effect/essentials_set/vali = list(CAT_ARMMOD, "Vali chemical enhancement set", 0,"black"),
```
Careful with making it free

Walarks replied:
People claim for it to be on the weak side currently. I added some usability optimizations and will see how it goes next testmerge.

Walarks replied:
First test people said it was on the weak side. I made some usability fixes and will see how it goes next testmerge

TiviPlus replied:
Aye hence why I said be careful

tgstation-server commented:
Your Fix/Feature pull request delta is currently below zero (-40). Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or helping to improve the codebase will raise this score.

fernandos13 commented:
VALI bug.. if you use 15u of bica in it

Then stab something

then you try to use any other bica to restock it
it will say no diference substance allowed

567Turtle commented:
I want you.
